### PR TITLE
Support cu_seqlens in unified SDPA

### PR DESCRIPTION
**Pull request by egilliam-nv**
## Summary

Exposes `cu_seqlens_*` (cumulative sequence length) parameters to SDPA users
as an alternative to per-batch `seq_len_q` / `seq_len_kv` tensors. Available
on the **unified** SDPA engine only, and requires cuDNN backend **>= 9.24**.

Constraints (enforced by the frontend):

- Forward-only (not plumbed for backward).
- Requires a padding mask.
- Mutually exclusive with `seq_len_q` / `seq_len_kv` tensors.
- `cu_seq_len_q` and `cu_seq_len_kv` must both be set or both unset.
- Compatible with paged K/V caches.

## Backend dependency

Backend MR (must ship first / be present in the targeted cuDNN release):
https://gitlab-master.nvidia.com/cudnn/cudnn/-/merge_requests/3063/

NVBug: https://nvbugspro.nvidia.com/bug/5833550

## Changes

- New `cu_seq_len_q` / `cu_seq_len_kv` inputs on `SDPA_attributes`.
- Validation in `DiagonalBandMaskNodeBase` (mutual exclusivity, paired set/unset).
- Validation in `verify_sdpa_support_surface_for_implementation`
  (`sdpa_support_surface.h`): paged K/V is allowed in combination with either
  `seq_len_*` or `cu_seq_len_*`. Standardized error code on the
  mutually-exclusive check to `INVALID_VALUE`.
- `UnifiedSDPANode`: gate on cuDNN >= 9.15 (shared with paged / `seq_len`)
  with a nested >= 9.24 check for `cu_seq_len_*`. Updated the
  "Paged attention attributes" comment block and error wording to cover
  all variable-length mechanisms. Added `CU_SEQ_LEN_Q` / `CU_SEQ_LEN_KV`
  to `collect_tensors_to_dump_node()`.
- New C++ sample: `samples/cpp/sdpa/fp16_fwd_with_cu_seq_len.cpp`
  (added to `samples/cpp/CMakeLists.txt`).

## Test plan

- [x] `samples/cpp/sdpa/fp16_fwd_with_cu_seq_len.cpp` builds cleanly.
- [x] `pytest -m L1 test/python/test_mhas_v2.py::test_sdpa_random_fwd_unified_L0 -n 8`
      passes against a cuDNN >= 9.24 backend.
- [x] `pytest -m L1 test/python/test_mhas_v2.py::test_sdpa_fwd_paged_unified_L0`
      passes all 128 iterations with no silent skips, confirming the paged +
      `cu_seq_len` validation fix in `sdpa_support_surface.h`.
- [x] CI on this PR.

## Summary by CodeRabbit

* **New Features**
  * Added support for per-batch cumulative sequence-length (cu_seq_len_q / cu_seq_len_kv) inputs across SDPA and diagonal-band-mask, including Python bindings and public API parameters.
  * Auto-promotion of 1D seq/offset tensors to the cuDNN-required 4D layout where needed.

* **Behavior / Validation**
  * Validation and implementation selection updated to enforce mutual exclusivity between seq_len and cu_seq_len inputs and to require both cu_seq_len tensors together; unified implementation gated by cuDNN version.

* **Samples**
  * Added a C++ sample demonstrating SDPA with cumulative sequence lengths.

* **Tests**
  * Expanded tests to exercise cu_seq_len modes and adjusted test utilities accordingly.

### CODE REVIEW

**Comment by egilliam-nv** [comment]
@cudnn-ci-bot run

File: include/cudnn_frontend/node/scaled_dot_product_flash_attention.h [hunk 1]
```diff
@@ -2379,12 +2422,46 @@ class UnifiedSDPANode : public SDPANodeBase<UnifiedSDPANode> {
                                                                &backend_seq_len_KV));
             }
 
+            // Optional cumulative-sequence-length tensors (mutually exclusive with
+            // SEQ_LEN_Q / SEQ_LEN_KV; pre-validated in validate_sdpa_support_surface()).
+            if (has_cu_seq_len_q() || has_cu_seq_len_kv()) {
```
Comment (Anerudhan) [inline include/cudnn_frontend/node/scaled_dot_product_flash_attention.h:2498]: Shouldnt this be in sdpa_support_surface ?

File: python/pygraph/sdpa.cpp [hunk 2]
```diff
@@ -595,6 +612,8 @@ PyGraph::sdpa_fp8(std::shared_ptr<cudnn_frontend::graph::Tensor_attributes>& q,
                                          use_padding_mask,
                                          seq_len_q,
                                          seq_len_kv,
+                                         cu_seq_len_q,
```
Comment (Anerudhan) [inline python/pygraph/sdpa.cpp:609]: Can you add additional argument at the end here and other places. Shouldnt be an issue here but for positional arg, could be breaking

File: include/cudnn_frontend/node/sdpa_support_surface.h [hunk 3]
```diff
@@ -79,13 +84,26 @@ SDPA_attributes::validate_sdpa_support_surface(const detail::Context& context,
...
+        (!padding_mask && !attention_score_modifier) &&
+            (has_seq_len_q || has_seq_len_kv || has_cu_seq_len_q || has_cu_seq_len_kv),
+        error_code_t::ATTRIBUTE_NOT_SET,
+        "seq_len_q/seq_len_kv (or cu_seq_len_q/cu_seq_len_kv) needs to be set only if padding mask is enabled.");
+
+    // Cumulative sequence length tensors must be set together, and are mutually
+    // exclusive with the (per-batch) seq_len tensors.
+    RETURN_CUDNN_FRONTEND_ERROR_IF(has_cu_seq_len_q != has_cu_seq_len_kv,
                                    error_code_t::ATTRIBUTE_NOT_SET,
-                                   "seq_len_q and seq_len_kv needs to be set only if padding mask is enabled.");
```
Comment (Anerudhan) [inline include/cudnn_frontend/node/sdpa_support_surface.h:88]: Also did not see case where both ragged and cuseqlen is given..

**Review by Anerudhan — COMMENTED** [review]
Can yo also add - support where the seqlens/ragged/cuseq are 1-D tensor and we allow auto-convert to 4-D.

File: include/cudnn_frontend/node/scaled_dot_product_flash_attention.h (hunk 1 above)
Comment (egilliam-nv) [reply include/cudnn_frontend/node/scaled_dot_product_flash_attention.h:2498]: There is already a guard in sdpa_support_surface, we only add CU_SEQ_LEN_* to the list of allowed inputs iff the backend version >= 9.24. The guard here in scaled_dot_product_flash_attention.h is primarily to prevent compiler errors when building against an old cuDNN that doesn't have the new attributes.

File: python/pygraph/sdpa.cpp (hunk 2 above)
Comment (egilliam-nv) [reply python/pygraph/sdpa.cpp:609]: Done.

File: include/cudnn_frontend/node/sdpa_support_surface.h (hunk 3 above)
Comment (egilliam-nv) [reply include/cudnn_frontend/node/sdpa_support_surface.h:88]: What's wrong with that case? We allow both ragged and (non cumulative) seqlen, that's how THD with the "partially packed layout" works (right?), so shouldn't we allow both ragged and cuseqlen?

**Comment by egilliam-nv** [comment]
@cudnn-ci-bot run

File: include/cudnn_frontend/node/sdpa_support_surface.h (hunk 3 above)
Comment (Anerudhan) [reply include/cudnn_frontend/node/sdpa_support_surface.h:88]: ragged and (non cumulative) seqlen 👍 
cuseqlen and (non cumulative) seqlen 👍 
Do we need :
ragged and  cuseqlen? How do we handle this? Error out?

**Review by Anerudhan — APPROVED** [review]

**Comment by Anerudhan** [comment]
Approved. 🚀 

Please rebase, squash and merge after pipeline is clean. 

Thanks @egilliam-nv 👍

File: include/cudnn_frontend/node/sdpa_support_surface.h (hunk 3 above)
Comment (egilliam-nv) [reply include/cudnn_frontend/node/sdpa_support_surface.h:88]: Those first two cases are different:

1. ragged and (non cumulative) seqlen together are **allowed**. In fact this is how non-packed THD works, because seqlen gives the actual length of each sequence while ragged gives the start position of each sequence.
2. cuseqlen and non-cumulative seqlen together are **not allowed** because they are two ways of specifying the same thing.

Given that #1 is allowed, I propose:
3. ragged and cuseqlen together should be **allowed**, since cuseqlen is just a different way of specifying seqlen.

Am I missing something?

**Comment by egilliam-nv** [comment]
@cudnn-ci-bot run

**Comment by egilliam-nv** [comment]
@cudnn-ci-bot run

**Comment by egilliam-nv** [comment]
@cudnn-ci-bot run

**Comment by egilliam-nv** [comment]
@cudnn-ci-bot run

File: test/python/sdpa/fp16.py [hunk 4]
```diff
@@ -628,8 +655,18 @@ def compute_and_compare_reference(cfg, allocs, tensors, diffs):
         k_ref = k_ref.to(cfg.data_type).float()
 
     dO_ref = dO_gpu.detach().float() if dO_gpu is not None else None
-    seq_len_q_ref = seq_len_q_gpu.flatten().detach() if seq_len_q_gpu is not None else None
-    seq_len_kv_ref = seq_len_kv_gpu.flatten().detach() if seq_len_kv_gpu is not None else None
+    # The reference accepts the per-batch seq_lens in either form: explicit seq_len_*
+    # tensors or the cumulative cu_seq_len_* tensors (recovered via consecutive diffs).
+    def _to_seq_len_ref(seq_len_gpu, cu_seq_len_gpu):
```
Comment (hwanseoc) [inline test/python/sdpa/fp16.py:660]: can you raise error if both are passed here?

File: test/python/sdpa/fp16.py [hunk 5]
```diff
@@ -76,6 +78,10 @@ def validate_config(cfg):
     if cfg.is_ragged:
         assert cfg.is_padding == True, "is_ragged=True and is_padding=False not allowed"
 
+    if cfg.is_cu_seq_len:
+        assert cfg.is_padding == True, "is_cu_seq_len=True requires is_padding=True"
+        assert cfg.is_train == False, "is_cu_seq_len=True is forward-only (cu_seq_len is not plumbed for backward)"
```
[the remark this answers is not in the document]
Comment (hwanseoc) [reply test/python/sdpa/fp16.py:83]: does not need to be waived in tester, already waived by NOT_SUPPORTED in the FE graph

**Review by hwanseoc — APPROVED** [review]
left 1 minor comment, LGTM
